rust/private/rustc.bzl

```diff
@@ -338,45 +351,97 @@ def rustc_compile_action(
+    command_env = _create_command_env(ctx, out_dir)
+    build_flags_expansion = (" $(cat '%s')" % build_info.flags.path) if build_info else ""
     # Handle that the binary name and crate name may be different.
-    # If a target name contains a - then cargo (and rules_rust) will generate a crate name with _ instead.
-    # Accordingly, rustc will generate a output file (executable, or rlib, or whatever) with _ not -.
-    # But when cargo puts a binary in the target/${config} directory, and sets environment variables like
-    # `CARGO_BIN_EXE_${binary_name}` it will use the - version not the _ version.
-    # So we rename the rustc-generated file (with _s) to have -s if needed.
+    #
+    # If a target name contains a - then cargo (and rules_rust) will generate a
+    # crate name with _ instead.  Accordingly, rustc will generate a output
+    # file (executable, or rlib, or whatever) with _ not -.  But when cargo
+    # puts a binary in the target/${config} directory, and sets environment
+    # variables like `CARGO_BIN_EXE_${binary_name}` it will use the - version
+    # not the _ version.  So we rename the rustc-generated file (with _s) to
+    # have -s if needed.
     maybe_rename = ""
     if crate_info.type == "bin":
         generated_file = crate_info.name
         if toolchain.target_arch == "wasm32":
             generated_file = generated_file + ".wasm"
-        src = "/".join([output_dir, generated_file])
+        src = "/".join([crate_info.output.dirname, generated_file])
         dst = crate_info.output.path
         if src != dst:
             maybe_rename = " && /bin/mv {src} {dst}".format(src=src, dst=dst)
-    command = '{}{}{}{} "$@" --remap-path-prefix="$(pwd)"=__bazel_redacted_pwd{}{}'.format(
-        ("export $(cat %s);" % build_info.rustc_env.path) if build_info else "",
-        manifest_dir_env,
-        out_dir_env,
+
+    return '{}{}{} "$@" --remap-path-prefix="$(pwd)"=__bazel_redacted_pwd{}{}'.format(
+        rustc_env_expansion,
+        command_env,
         toolchain.rustc.path,
-        (" $(cat '%s')" % build_info.flags.path) if build_info else "",
+        build_flags_expansion,
         maybe_rename,
     )
 
+def rustc_compile_action(
+        ctx,
+        toolchain,
+        crate_info,
+        output_hash = None,
+        rust_flags = []):
+    """
+    Constructs the rustc command used to build the current target.
+
+    Returns:
+      List[Provider]: A list of the following providers:
+                     - CrateInfo: info for the crate we just built; same as `crate_info` parameter.
+                     - DepInfo: The transitive dependencies of this crate.
+                     - DefaultInfo: The output file for this crate, and its runfiles.
+    """
+    dep_info, build_info = collect_deps(
+        crate_info.deps,
+        crate_info.aliases,
+        toolchain,
+    )
+
+    compile_inputs, out_dir = _rustc_compile_inputs(
```

Comment:
I think each of these subcommands should start w/ verbs, eg.

`collect_deps`
`collect_inputs`
`construct_arguments` etc

Reply:
Done